mockgen/import_mode.go

```diff
@@ -0,0 +1,376 @@
+
+	pkg := pkgs[0]
+	searchedPkg, ok := pkg.Imports[packageName]
+	if !ok {
+		return nil, fmt.Errorf("package %s not found", packageName)
+	}
+
+	return searchedPkg, nil
+}
+
+func (p *importModeParser) generateSalt() string {
+	var numbers = []byte("0123456789")
+
+	result := make([]byte, 8)
+	for i := range result {
+		result[i] = numbers[rand.Intn(len(numbers))]
+	}
+
+	return string(result)
+}
+
+func (p *importModeParser) extractInterfacesFromPackage(pkg *packages.Package, ifaces []string) ([]*model.Interface, error) {
+	interfaces := make([]*model.Interface, len(ifaces))
+	for i, iface := range ifaces {
+		obj := pkg.Types.Scope().Lookup(iface)
+		if obj == nil {
+			return nil, fmt.Errorf("interface %s does not exist", iface)
+		}
+
+		modelIface, err := p.parseInterface(obj)
+		if err != nil {
+			return nil, newParseTypeError("parse interface", obj.String(), err)
+		}
+
+		interfaces[i] = modelIface
+	}
+
+	return interfaces, nil
+}
+
+func (p *importModeParser) parseInterface(obj types.Object) (*model.Interface, error) {
+	named, ok := obj.Type().(*types.Named)
+	if !ok {
+		return nil, fmt.Errorf("%s is not an interface. it is a %s", obj.Name(), obj.Type().Underlying().String())
+	}
+
+	iface, ok := named.Underlying().(*types.Interface)
+	if !ok {
+		return nil, fmt.Errorf("%s is not an interface. it is a %s", obj.Name(), obj.Type().Underlying().String())
+	}
+
+	if p.isConstraint(iface) {
+		return nil, fmt.Errorf("interface %s is a constraint", obj.Name())
+	}
+
+	methods := make([]*model.Method, iface.NumMethods())
+	for i := 0; i < iface.NumMethods(); i++ {
+		method := iface.Method(i)
+		typedMethod, ok := method.Type().(*types.Signature)
+		if !ok {
+			return nil, fmt.Errorf("method %s is not a signature", method.Name())
+		}
+
+		modelFunc, err := p.parseFunc(typedMethod)
+		if err != nil {
+			return nil, newParseTypeError("parse method", typedMethod.String(), err)
+		}
+
+		methods[i] = &model.Method{
+			Name:     method.Name(),
+			In:       modelFunc.In,
+			Out:      modelFunc.Out,
+			Variadic: modelFunc.Variadic,
+		}
+	}
+
+	var typeParams []*model.Parameter
+	if named.TypeParams() != nil {
+		typeParams = make([]*model.Parameter, named.TypeParams().Len())
+		for i := 0; i < named.TypeParams().Len(); i++ {
+			param := named.TypeParams().At(i)
+			typeParam, err := p.parseConstraint(param)
+			if err != nil {
+				return nil, newParseTypeError("parse type parameter", param.String(), err)
+			}
+
+			typeParams[i] = &model.Parameter{Name: param.Obj().Name(), Type: typeParam}
+		}
+	}
+
+	return &model.Interface{Name: obj.Name(), Methods: methods, TypeParams: typeParams}, nil
+}
+
+func (o *importModeParser) isConstraint(t *types.Interface) bool {
+	for i := 0; i < t.NumEmbeddeds(); i++ {
+		embed := t.EmbeddedType(i)
+		if _, ok := embed.Underlying().(*types.Interface); !ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (p *importModeParser) parseType(t types.Type) (model.Type, error) {
+	switch t := t.(type) {
```

Comment:
In a similar vein as my other comments, we need to handle `*types.Alias` here. Anywhere along the type traversal can be an alias, including function parameters, return values, slice elements, etc.

I can think of two ways to handle this:
* We can add explicit *types.Alias representation in `model` via a new `Alias` type, that will simply point to the underlying type. This follows upstream Go's model of introducing `Alias` to `go/types`.
* We can diligently unalias types we encounter via `types.Unalias` while parsing types and converting to gomock's model representation. For example, we'd add a case in this switch to passthrough aliases. Something like:
```
case *types.Alias:
     return p.parseType(types.Unalias(t))
```

I don't have a strong preference, but until there's a reason for mock to provide special treatment & explicit representation to aliases, I'm fine with the latter.

Reply:
`types.Unalias` will work, but as a result we will get the true type instead of the alias.
My idea is that `types.Alias` is the same as `types.Named` in our case. What we need from alias? Only his name.
Therefore, we can combine both types under a single interface.
Like that:
```go
type namedType interface {
	Obj() *types.TypeName
	TypeArgs() *types.TypeList
}

func (p *importModeParser) parseType(t types.Type) (model.Type, error) {
...
	case *types.Named, *types.Alias:
		named := t.(namedType)
		var pkg string
		if named.Obj().Pkg() != nil {
			pkg = named.Obj().Pkg().Path()
		}
	
		if named.TypeArgs() == nil {
			return &model.NamedType{
				Package: pkg,
				Type:    named.Obj().Name(),
			}, nil
		}
	
		typeParams := &model.TypeParametersType{TypeParameters: make([]model.Type, named.TypeArgs().Len())}
		for i := 0; i < named.TypeArgs().Len(); i++ {
			typeParam := named.TypeArgs().At(i)
			typedParam, err := p.parseType(typeParam)
			if err != nil {
				return nil, newParseTypeError("parse type parameter", typeParam.String(), err)
			}
	
			typeParams.TypeParameters[i] = typedParam
		}
	
		return &model.NamedType{
			Package:    pkg,
			Type:       named.Obj().Name(),
			TypeParams: typeParams,
		}, nil
...
}
```

Reply:
I think that would be a nice/convenient solution, but the problem with that approach that I see is that Aliases aren't always to named types. You can have an alias to `int`, or `map[string]struct{}`, or even an alias to an anonymous struct definition, e.g., `type MyStruct = struct { ... }`. In these cases, we would be returning a `model.NamedType` when it's technically not appropriate.

I think it's probably better (and more in the spirit of recursive analysis) to defer to how the underlying type would be parsed by calling `p.parseType(types.Unalias(t))`.

Reply:
I _think_ this sounds reasonable - I can't immediately think of a case where we'd need the underlying type's information since it seems like we only need to refer to its name.